### ISSUE TITLE
Magnetar Stacking Damage Penalty

### ISSUE THREAD

**Issue by PaladinOne** [pyfa-org/Pyfa#2009]
another old throwback bug, this one back to #159...

A friend and I were testing Dreadnought fits (Naglfar and Phoenix), with a C5 Magnetar effect projected onto them. We noticed that the displayed damage numbers were a bit below where they should be. Examination of the guns' Affected By tab, and subsequently looking at /eos/effects.py indicates that the projected Magnetar Effect was being stacking penalized, which it should not be.

We then tested this explicitly in game using a Rattlesnake, with one Gecko, and 2 Drone Damage Amplifiers, in a C5 Magnetar. Strangely, for this test, PYFA gave the _correct_ values, with no stacking penalty from the Magnetar effect. So the bug seems to specifically be affecting turrets/launchers, and not Drones.

v2.9.4, Windows 10

**Comment by DarkFenX** [comment]
How it should work according to theory, after looking at data:

- Guns: regular stacking penalty for everything (magnetar uses postMul, damage mods use postMul)
- Drones and fighters: damage mods are in one stacking chain, magnetar is in another (magnetar uses postMul, DDA uses postPercent)
- Missiles: damage mods are in one stacking chain, magnetar is in another (magnetar uses postMul, BCS uses preMul and affects missile damage via proxy on-character attribute)

It should be stacking penalized, but just should be in another stacking penalty chain.

**Comment by DarkFenX** [comment]
Now, how it is implemented in pyfa. Missiles:

- missileDMGBonus: stacking penalized, default stacking chain
- systemDamageEmMissiles: stacking penalized, postMul stacking chain

Missiles should be correctly calculated. Please share full fit and skills, the value you have in pyfa and the value you see in game. Strip of irrelevant modules (those which do not affect missile damage) if you prefer to conceal the fit.

Guns:
- projectileWeaponDamageMultiply: stacking penalized, default stacking chain
- systemDamageMultiplierGunnery: stacking penalized, default stacking chain

Those should also show correct values. If after double-check they do not - do the same for naglfar.

Drones:
- moduleBonusDroneDamageAmplifier: stacking penalized, default stacking penalty chain
- systemDamageDrones: stacking penalized, postMul stacking penalty chain
- systemDamageFighters: stacking penalized, postMul stacking penalty chain

Those should also be correct

**Comment by DarkFenX** [comment]
C3 magnetar.

Gun test:
```
[Retribution, #Refraction]

Corpii A-Type Small Armor Repairer
Nanofiber Internal Structure II
Nanofiber Internal Structure II
Heat Sink II
Imperial Navy Heat Sink

Coreli A-Type 5MN Microwarpdrive
Shadow Serpentis Warp Disruptor

Small Focused Beam Laser II, True Sanshas Xray S
Small Focused Beam Laser II, True Sanshas Xray S
[Empty High slot]
Small Focused Beam Laser II, True Sanshas Xray S
Small Focused Beam Laser II, True Sanshas Xray S

Small Energy Locus Coordinator II
Small Energy Locus Coordinator II


High-grade Snake Alpha
High-grade Snake Beta
High-grade Snake Gamma
High-grade Snake Delta
High-grade Snake Epsilon
High-grade Snake Omega
Eifyr and Co. 'Rogue' Evasive Maneuvering EM-705
Zor's Custom Navigation Hyper-Link
Eifyr and Co. 'Gunslinger' Surgical Strike SS-905
```
In-game: 20.67x damage modifier
Pyfa: 20.6697x damage modifier
DPS readout coincides as well.


Missile test:
```
[Garmur, #Hornet]

Caldari Navy Ballistic Control System
Caldari Navy Ballistic Control System
Nanofiber Internal Structure II

Coreli A-Type 5MN Microwarpdrive
Republic Fleet Medium Shield Extender
Dark Blood Warp Scrambler
Republic Fleet Warp Disruptor

Dread Guristas Light Missile Launcher, Caldari Navy Scourge Light Missile
Dread Guristas Light Missile Launcher, Caldari Navy Scourge Light Missile
Dread Guristas Light Missile Launcher, Caldari Navy Scourge Light Missile

Small Ionic Field Projector II
Small Ancillary Current Router I
Small Polycarbon Engine Housing II


High-grade Snake Alpha
High-grade Snake Beta
High-grade Snake Gamma
High-grade Snake Delta
High-grade Snake Epsilon
High-grade Snake Omega
Eifyr and Co. 'Rogue' Evasive Maneuvering EM-705
Zor's Custom Navigation Hyper-Link
Zainou 'Snapshot' Light Missiles LM-905
Zainou 'Deadeye' Rapid Launch RL-1005
```
In-game: 603.06 kin damage on missile, 276.1 dps
Pyfa: 603.0566 kin damage, 276.1 dps

Drone test:
```
[Astero, Acolyte snake]

Damage Control II
Drone Damage Amplifier II
Drone Damage Amplifier II
Micro Auxiliary Power Core I

5MN Quad LiF Restrained Microwarpdrive
Republic Fleet Medium Shield Extender
Fleeting Compact Stasis Webifier
Caldari Navy Warp Scrambler

Covert Ops Cloaking Device II
Small Focused Pulse Laser II, Scorch S

Small Core Defense Field Extender II
Small Core Defense Field Extender II
Small Core Defense Field Extender II


Acolyte II x5
Hobgoblin II x5
Hobgoblin II x5


High-grade Snake Alpha
High-grade Snake Beta
High-grade Snake Gamma
High-grade Snake Delta
High-grade Snake Epsilon
High-grade Snake Omega
Eifyr and Co. 'Rogue' Evasive Maneuvering EM-705
Zor's Custom Navigation Hyper-Link
Eifyr and Co. 'Gunslinger' Surgical Strike SS-905
Inherent Implants 'Lancer' Large Energy Turret LE-1005
```
In-game: 8.883x hob dmg mult, 263.3 dps
Pyfa: 8.8827x hob dmg mult, 263.3 dps

Conclusion is that damage mod vs magnetar stacking penalty seems to be working as intended, but there might be some shenanigans with siege module.

**Comment by DarkFenX** [comment]
Another test, this time on sisi. Revelation, t1 guns, t2 siege, 3x t2 heat sinks, c2 magnetar, cap gun skill 4.

EVE: 23.735x dmg mult, 2069.5 dps
Pyfa: 23.734634 dmg mult, 2069.5 dps

So, looks like at least for HAW guns stacking penalty is working like intended.

**Comment by DarkFenX** [comment]
While working on the issue, I discovered the following:

1) Last 2 releases phoenix had quad thermal bonus in pyfa, instead of 4 separate bonuses for each damage type. This is fixed now.
2) There seems to be some bug in EVE. My phoenix with t2 RTLs has less RoF than it should have. I have equivalent of all V character for phoenix. In game, it has 39.57 cycle time. However, it should be 51.722×0.9×0.9×0.85 = 35.610597. If you take one of 0.9's away - it becomes 39.56733 which is what we see in game.

I believe that when CCP added requiredSkill4/5/6, they forgot to include them into skill requirement checker used in dogma. requiredSkill4 for t2 RTL is missile launcher operation skill which provides 10% rof. For t2 HAW 800mms requiredSkill4 is gunnery which again provides 10% rof - and i suppose that it will be missing as well, compared to expected numbers.

I filled bug report in CCP bug tracker, but it's either I am tired and fucked up, or bug has to be confirmed by CCP - I have no ID of it still.

**Comment by DarkFenX** [comment]
Phoenix (t2 siege, 3x t2 BCS, 2x t1 xl cruise launchers with t1 em missiles, 1x t2 rapid torpedo launcher with rage thermal missiles), capital launcher operation skills are 1, the rest are 5.

Rapid Torp Launchers:
- Cycle time: 3.03 in game, pyfa: 2.951823 with stacking penalty from siege, 2.722776 without (exactly 10% less due to a bug above)
- Damage: 5705.4 in game, 5705.4 in game

XL cruise launchers:
- Cycle time: 20.00 in game, 20.00064656 in pyfa
- Damage: 37441.68 in game, 37441.6835 in pyfa

So by the looks of it CCP fixed siege HAW bonus - it is no longer stacking penalized with damage mods. This issue has nothing to do with Magnetar effect.

**Comment by DarkFenX** [comment]
Naglfar (t2 siege, 3x t2 gyrostabs, 2x t1 haws, 1x t1 3500mm), capital gun operation skills are 1, minmatar dread skill is 1.

800mms:
- Damage mult: 11.198 in-game, 11.19828 in pyfa
- Cycle time: 2.09 in-game, 2.0919 in pyfa

3500mms:
- Damage mult: 209.096 in-game, 209.0942 in pyfa
- Cycle time: 15.69 in-game, 15.6895

So, for Naglfar pyfa shows correct numbers with all the modifiers (damage mods, siege, magnetar).

**Comment by DarkFenX** [comment]
Looks like it was an omission on our part. Damage mods use postMul for modification, siege module uses postPercent. Fixed in cecf5d7e314cfe87320521961f493db58dceeb5e.

So, to summarize:

- quad-thermal damage bonus on phoenix was fixed
- stacking penalty for phoenix HAW from siege was fixed
- regarding original complaint - magnetar IS stacking penalized and it should be. All system-wide effects are penalized, they just might use another operator for modification (and penalizations are applied within scope of single operator).
- regarding missing 10% rof bonus on t2 RTL on sisi - bug is filled (EBR-180798), it's up to CCP to fix it. Pinging @tsuthers-ccpgames for visibility (see this [link](https://github.com/pyfa-org/Pyfa/issues/2009#issuecomment-520271660) for more info).

**Comment by DarkFenX** [comment]
@PaladinOne if you can give exact numbers - please post them here still, maybe I've missed something.
